File: 0121. Best Time to Buy and Sell Stock.cpp
//Time Complexity: O(n)
//Space Complexity: O(n)
//Approach: Prefix Min and Suffix Max
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();

        vector<int> minPref;
        int M = 2147483647;
        for(int i : prices){
            M = min(i, M);
            minPref.push_back(M);
        }
        
        vector<int> maxSuf;
        M = 0;
        for(int i = 0; i < n; i++){
            M = max(prices[n-i-1], M);
            maxSuf.push_back(M);
        }

        int res = 0;
        for(int i = 0; i < n; i++){
            res = max(maxSuf[n-i-1] - minPref[i], res);
        }
        return res;
    }
};

File: 0121. Best Time to Buy and Sell Stock.py
# Time Complexity: O(n)
# Space Complexity: O(n)
# Approach: Prefix Min, Suffix Max

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0
        
        m = [prices[0]]
        for p in prices[1:]:
            m.append(min(p, m[-1]))
        
        M = [0 for i in prices]
        M[-1] = prices[-1]
        for p in range(len(prices)-2, -1, -1):
            M[p] = max(prices[p], M[p+1])

        res = -float(inf)
        for i in range(len(prices)):
            res = max(res, M[i] - m[i])
        return res
